src/router.rs

```diff
@@ -1,19 +1,38 @@
-use crate::cir::{CirProgram, Translate};
+use crate::ast::Value;
+use crate::cir::{CirInstruction, CirProgram, Translate};
 use crate::context::{Context, Match};
 use crate::interpreter::Execute;
 use crate::parser::parse;
 use crate::schema::Schema;
 use crate::semantics::{FieldCounter, Validate};
+use regex::Regex;
 use std::collections::{BTreeMap, HashMap};
+use std::rc::Rc;
 use uuid::Uuid;
 
+// use crate::ast::{Expression, LogicalExpression, Value};
+
 #[derive(PartialEq, Eq, PartialOrd, Ord)]
 struct MatcherKey(usize, Uuid);
 
 pub struct Router<'a> {
     schema: &'a Schema,
     matchers: BTreeMap<MatcherKey, CirProgram>,
     pub fields: HashMap<String, usize>,
+    pub regex_cache: HashMap<String, Rc<Regex>>,
+}
+
+fn release_cache(cir: &CirProgram, router: &mut Router) {
+    cir.instructions.iter().for_each(|instruction| {
```

Comment:
Since instructions are now just in an array I can simply iterate through it and release regexes when I encounter them. Previously I had to traverse the tree of AST.